### ISSUE TITLE
Further exploration required - OOB sample for estimating residual distribution?

### ISSUE THREAD

**Issue by markjrieke** [markjrieke/workboots#21]
Currently, `predict_boots()` uses the residuals from the training set to estimate the prediction interval. This lines up wiht the methodology from *Boostrap Methods and their Application*, but as Max points out in #18 this may give unrealistic results for over-fit models. If the training error goes to 0, `predict_boots()` will give a prediction interval of +/- 0. Some to-do's:

- [x] branch for testing this out
- [x] modify `predict_boots()` to sample residuals/use the distribution from the residuals on the OOB sample
- [x] compare linear model's theoretical interval (lm) against linear model's bootstrap output (OOB sample)
- [x] compare linear model's bootstrap output (training sample) against linear model's bootstrap output (OOB sample)
- [x] compare other model types' training sample against OOB sample

**Comment by markjrieke** [comment]
[This CV discussion](https://stats.stackexchange.com/questions/96739/what-is-the-632-rule-in-bootstrapping) gives a good overview of the 632 estimate. The 632 estimate gives an approximate error for the model by blending training/oob error but can be faulty when overfitting (i.e., training error -> 0). The 632+ estimate may be able to account for this. 

The 632 estimate, however, only gives a single +/- error val, whereas predict boots is drawing from the residual distribution (or normal approx. of the residuals) --- so I'm not sure if there's a way to incorporate without *losing* some information from the residual distribution...

**Comment by markjrieke** [comment]
lm comparison - see third checklist item above (not rerunning a reprex, but code + output here):

```r
library(tidyverse)
library(tidymodels)

data("penguins")
penguins_mod <- 
  penguins %>% 
  drop_na()

set.seed(999)
penguins_split <- initial_split(penguins_mod)
penguins_train <- training(penguins_split)
penguins_test <- testing(penguins_split)

penguins_wf <-
  workflow() %>%
  add_recipe(recipe(body_mass_g ~ flipper_length_mm, data = penguins_train)) %>%
  add_model(linear_reg())

set.seed(888)
penguins_boots <- 
  penguins_wf %>%
  predict_boots(
    n = 2000, 
    training_data = penguins_train, 
    new_data = penguins_test
  )

set.seed(777)
penguins_lm <- 
  lm(body_mass_g ~ flipper_length_mm, data = penguins_train)

penguins_lm %>%
  predict(penguins_test, interval = "predict") %>%
  as_tibble() %>%
  bind_cols(penguins_boots %>% summarise_predictions()) %>%
  bind_cols(penguins_test) %>%
  ggplot(aes(x = flipper_length_mm)) +
  geom_point(aes(y = body_mass_g),
             alpha = 0.25) +
  geom_line(aes(y = fit),
            size = 1) +
  geom_ribbon(aes(ymin = lwr,
                  ymax = upr),
              alpha = 0.15) +
  geom_line(aes(y = .pred),
            size = 1,
            color = "blue") +
  geom_ribbon(aes(ymin = .pred_lower,
                  ymax = .pred_upper),
              alpha = 0.15,
              fill = "blue")
```
<img width="596" alt="image" src="https://user-images.githubusercontent.com/31866288/158234334-5547115a-79f8-4227-bbeb-fb32d336c949.png">

**Comment by markjrieke** [comment]
lm train vs. oob residuals (addresses 4th checklist item above); again not re-running for the reprex but here is the code + output

```r
library(tidyverse)
library(tidymodels)

data("penguins")
penguins_mod <- 
  penguins %>% 
  drop_na()

set.seed(999)
penguins_split <- initial_split(penguins_mod)
penguins_train <- training(penguins_split)
penguins_test <- testing(penguins_split)

penguins_wf <-
  workflow() %>%
  add_recipe(recipe(body_mass_g ~ flipper_length_mm, data = penguins_train)) %>%
  add_model(linear_reg())

set.seed(888)
penguins_boots_oob <- 
  penguins_wf %>%
  predict_boots(
    n = 2000, 
    training_data = penguins_train, 
    new_data = penguins_test,
    oob = TRUE
  )

set.seed(777)
penguins_boots_trn <-
  penguins_wf %>%
  predict_boots(
    n = 2000,
    training_data = penguins_train,
    new_data = penguins_test,
    oob = FALSE
  )

penguins_boots_oob %>%
  summarise_predictions() %>%
  select(oob.pred_lower = .pred_lower,
         oob.pred = .pred,
         oob.pred_upper = .pred_upper) %>%
  bind_cols(penguins_boots_trn) %>%
  summarise_predictions() %>%
  select(starts_with("oob"),
         trn.pred_lower = .pred_lower,
         trn.pred = .pred,
         trn.pred_upper = .pred_upper) %>%
  bind_cols(penguins_test) %>%
  ggplot(aes(x = flipper_length_mm)) +
  geom_point(aes(y = body_mass_g),
             alpha = 0.25) +
  geom_line(aes(y = oob.pred),
            size = 1,
            color = "blue",
            alpha = 0.5) +
  geom_line(aes(y = trn.pred),
            size = 1,
            color = "green",
            alpha = 0.5) +
  geom_ribbon(aes(ymin = oob.pred_lower,
                  ymax = oob.pred_upper),
              fill = "blue",
              alpha = 0.15) +
  geom_ribbon(aes(ymin = trn.pred_lower,
                  ymax = trn.pred_upper),
              fill = "green",
              alpha = 0.15)
```
<img width="479" alt="image" src="https://user-images.githubusercontent.com/31866288/158241681-f632b146-4692-43f2-8efc-36fd017adb1e.png">

**Comment by markjrieke** [comment]
As expected, using the oob residuals results in a wider prediction interval than using training residuals; if this were way overfit, the training spread would potentially be 0.

```r
library(tidyverse)
library(tidymodels)

data("penguins")
penguins_mod <- 
  penguins %>% 
  drop_na()

set.seed(999)
penguins_split <- initial_split(penguins_mod)
penguins_train <- training(penguins_split)
penguins_test <- testing(penguins_split)

penguins_wf <-
  workflow() %>%
  add_recipe(recipe(body_mass_g ~ flipper_length_mm, data = penguins_train)) %>%
  add_model(boost_tree("regression"))

set.seed(888)
penguins_boots_oob <- 
  penguins_wf %>%
  predict_boots(
    n = 2000, 
    training_data = penguins_train, 
    new_data = penguins_test,
    oob = TRUE
  )

set.seed(777)
penguins_boots_trn <-
  penguins_wf %>%
  predict_boots(
    n = 2000,
    training_data = penguins_train,
    new_data = penguins_test,
    oob = FALSE
  )

penguins_boots_oob %>%
  summarise_predictions() %>%
  select(oob.pred_lower = .pred_lower,
         oob.pred = .pred,
         oob.pred_upper = .pred_upper) %>%
  bind_cols(penguins_boots_trn) %>%
  summarise_predictions() %>%
  select(starts_with("oob"),
         trn.pred_lower = .pred_lower,
         trn.pred = .pred,
         trn.pred_upper = .pred_upper) %>%
  mutate(oob.spread = oob.pred_upper - oob.pred_lower,
         trn.spread = trn.pred_upper - trn.pred_lower) %>%
  bind_cols(penguins_test) %>%
  pivot_longer(ends_with("spread"),
               names_to = "resid_type",
               values_to = "spread") %>%
  ggplot(aes(x = body_mass_g,
             y = spread,
             color = resid_type)) +
  geom_point() +
  geom_smooth(se = FALSE)
```
<img width="595" alt="image" src="https://user-images.githubusercontent.com/31866288/158246869-2b0da490-c0f9-4743-909b-8df5db2934df.png">

**Comment by markjrieke** [comment]
addressed by #22
